### DISCUSSION TITLE
How do I troubleshoot third-party authentication errors?

### DISCUSSION THREAD

**Discussion by MarkMYoung** [alexeybusygin/ShippingRates#58] [Q&A]
An intern developed a proof-of-concept (POC) using this library.  When my timeline caught up to the work, I took that POC with a `IRateAdjuster` and incorporated it into development.  For no apparent reason, the USPS `userId` we were using was authenticating then it just stopped authenticating, yet it produced no errors.  Additionally, I added FedEx credentials (which I got from our FedEx account), but they produce a vague error (it doesn't even indicate that it's from FedEx versus DHL, etc.).   I don't know how there is an authentication error since I can log in and verify that my credentials are correct (I am specifying `AccountNumber`, `Key`, `MeterNumber`, and `Password`).

I am using `GetRatesAsync(Address originAddress, Address destinationAddress, Package package, ShipmentOptions options = null)`.  Again, USPS produces no error.  The FedEx error:

```JSON
"Errors":
[
	{
		"Description": "Authentication Failed",
		"HelpContext": null,
		"HelpFile": null,
		"Number": "1000",
		"Source": "prof"
	}
]
```

Any ideas on next steps for troubleshooting?

Thanks in advance,
Mark

**Comment by alexeybusygin** [comment]
Hi Mark, I can reproduce this error if the key, password, or account number is incorrect. If you use an incorrect meter number or dev credentials against a live endpoint (and vice versa), you will get "Meter number is missing or invalid." I suggest double-checking the key, password, and account number.

**Reply by MarkMYoung** [reply]
The FedEx credentials were all correct.  We have it working now with both `UseProduction` true and false.  We had to convert the
`Shipment shipment = await rateManager.GetRatesAsync(originAddress, destinationAddress, package);`
to
`Shipment shipment = rateManager.GetRatesAsync(originAddress, destinationAddress, package).Result;`

After that, it went from working 5--10% of the time to 85--90% of the time.

**Reply by stephendglasgow** [reply]
@alexeybusygin - I wanted to follow up on Mark's issue above.  I work with him and am helping to troubleshoot some issues we're seeing with sporadic Authentication Failed errors coming from FedEx when we try to get shipping rates, this is using their Prod environment (so useProduction is set to true).

I actually just got off the phone with FedEx tech support and explained how we will, pretty consistently, get an Authentication Failed error when making one of these requests, but we'll immediately try again and it works fine.  They explained to me that they are currently in a transition to RESTful API's and are continuously making updates to their Prod and Test environments, resulting in the environments being unstable so that behavior is expected.  They said this is expected for the next couple of weeks or around the end of the month (Aug 2024).

So my question for you, just to make sure our bases are covered, is whether that transition will cause any issues with us using this library or if it should continue to work as usual.  We are currently using version 1.1.10.204.  I noticed there's a few newer versions since then, and a note on the GitHub readme about version 2.1.0 having some changes to fix issues with UPS due to changes they made on their end, but we're not currently using UPS so it doesn't seem like we'd need to update the library.

Overall, I'm just trying to get ahead of this so if we do need to update our ShippingRates library version to support the FedEx changes, or if the library will stop working when those changes are released for FedEx, then we have some time to make whatever changes are necessary.

I forgot to mention, when I asked FedEx where we could see the latest updates/news about this transition, they directed me to this page that has a banner at the top which is supposed to show the latest info:  https://www.fedex.com/en-us/developer.html

Thank you for your help

**Reply by alexeybusygin** [reply]
So far, FedEx has announced disabling some of its services ("FedEx Web Services Tracking, Address Validation, and Validate Postal Codes WSDLS will be disabled on August 31, 2024"), but the Quote Rates service was not mentioned. However, we can expect that they will shut down the Quote Rates service too, at some point. So, later this year, we are going to implement support for FedEx REST APIs as well. And while FedEx supports both services we will support both services too.

I don't think there have been any changes related to FedEx authentication since version 1.1.10.204, but we are fixing different bugs sometimes, and upgrading to the latest version seems like a good idea to me 😄 

From our experience, FedEx APIs can be not so stable. But having a lot of Authentication Failed errors is unusual. 

You may also want to try sync execution, so instead of calling:
`Shipment shipment = rateManager.GetRatesAsync(originAddress, destinationAddress, package).Result;`
call:
`Shipment shipment = rateManager.GetRates(originAddress, destinationAddress, package);`
